### PR TITLE
PERL-918 Update retryable writes spec

### PR DESCRIPTION
**Pull request by TBSliver**
Updated retryable-writes spec tests as the appeared at mongodb/specifications@b7c53fc03a35d8a7ed586ddb8c9cfd95eb6e412b

### CODE REVIEW

File: lib/MongoDB/MongoClient.pm [hunk 1]
```diff
@@ -1624,7 +1624,8 @@ sub send_retryable_write_op {
         my $retry_link = $self->{_topology}->get_writable_link;
 
         # Rare chance that the new link is not retryable
```
Comment (xdg) [inline lib/MongoDB/MongoClient.pm]: Comment is now mildly confusing given extra error check.

File: lib/MongoDB/Role/_SessionSupport.pm [hunk 2]
```diff
@@ -92,6 +92,20 @@ sub _update_operation_time {
     return;
 }
 
+# Certain errors have to happen as soon as possible, such as write concern
+# errors in a retryable write. This has to be seperate to the other functions
+# due to not all result objects having the base response inside, so cannot be
+# used to parse operationTime or $clusterTime
+sub _assert_session_errors {
```
Comment (xdg) [inline lib/MongoDB/Role/_SessionSupport.pm]: I'm a little nervous about this, but if all existing tests pass (particularly for bulk operations) then I'm okay with it.

File: lib/MongoDB/MongoClient.pm (hunk 1 above)
Comment (TBSliver) [reply lib/MongoDB/MongoClient.pm]: Oh oops I did have a comment there about the extra check, must have got caught with a debugging cleanup...

File: lib/MongoDB/Role/_SessionSupport.pm (hunk 2 above)
Comment (TBSliver) [reply lib/MongoDB/Role/_SessionSupport.pm]: Yea this was the least complicated place to add this assertion - I couldn't see any information about the comment in `MongoDB::Op::_BulkWrite->execute` of `write concern errors are thrown only for the entire batch` in the specs for bulk operations, and as retryable writes can happen on certain write concern errors it must be done inside the retry loop.

The other single ops do throw WriteConcernErrors further down the stack, but only on the return value of the `send_*write_op` which is outside the retryable writes block - could maybe move that assertion up the call stack slightly to be inside the execute block? If commandResult is also modified to check for write concern errors in the assert step, that would get rid of this extra assertion? Would mean that bulk writes would throw write concern errors earlier, and as I said I couldnt find if that is a spec decision or not...

**Comment by TBSliver** [comment]
As far as I can tell, `MongoDB::NetworkError` should have `__is_retryable_error` be true, as it extends `MongoDB::ConnectionError` - unless I've missed something in the inheritance? Have updated the comment to be a bit clearer in the retry logic and re-arranged it slightly, as I noticed a possible issue...

**Comment by xdg** [comment]
Closing this PR as the #163 includes it.

**Pull request closed without merging**